neqo-transport/src/events.rs

```diff
@@ -99,23 +111,131 @@ impl ConnectionEvents {
     }
 
     pub fn connection_state_change(&self, state: State) {
+        // If closing, existing events no longer relevant.
+        match state {
+            State::Closing { .. } | State::Closed(_) => self.events.borrow_mut().clear(),
+            _ => (),
+        }
         self.insert(ConnectionEvent::StateChange(state));
     }
 
     pub fn client_0rtt_rejected(&self) {
+        // If 0rtt rejected, must start over and existing events are no longer
+        // relevant.
         self.events.borrow_mut().clear();
         self.insert(ConnectionEvent::ZeroRttRejected);
     }
 
     pub fn events(&self) -> impl Iterator<Item = ConnectionEvent> {
-        self.events.replace(BTreeSet::new()).into_iter()
+        self.events.replace(VecDeque::new()).into_iter()
+    }
+
+    pub fn has_events(&self) -> bool {
+        !self.events.borrow().is_empty()
     }
 
+    pub fn next_event(&self) -> Option<ConnectionEvent> {
+        self.events.borrow_mut().pop_front()
+    }
+
+    #[allow(clippy::block_in_if_condition_stmt)]
     fn insert(&self, event: ConnectionEvent) {
-        self.events.borrow_mut().insert(event);
+        let mut q = self.events.borrow_mut();
+
+        // Special-case two enums that are not strictly PartialEq equal but that
+        // we wish to avoid inserting duplicates.
+        if match &event {
```

Comment:
I think that you should factor out the "already_present" check into another function.  Then you can remove the `#[allow`